bundledatasrc.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package charm
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/juju/errors"
+	"gopkg.in/yaml.v2"
+)
+
+type fieldPresenseMap map[interface{}]interface{}
+
+func (fpm fieldPresenseMap) fieldPresent(fieldName string) bool {
+	_, exists := fpm[fieldName]
+	return exists
+}
+
+func (fpm fieldPresenseMap) forField(fieldName string) fieldPresenseMap {
+	v, exists := fpm[fieldName]
+	if !exists {
+		return nil
+	}
+
+	asMap, valid := v.(fieldPresenseMap)
+	if !valid {
+		panic(errors.Errorf("field map entry %q does not point to a nested field presense map", fieldName))
+	}
+	return asMap
+}
+
+// bundleDataPart combines a parsed BundleData instance with a nested map that
+// can be used to discriminate between fields that are missing from the data
+// and those that are present but defined to be empty.
+type bundleDataPart struct {
+	data        *BundleData
+	presenseMap fieldPresenseMap
+}
+
+// BundleDataSource is implemented by types that can parse bundle data into a
+// list of composable parts.
+type BundleDataSource interface {
+	read() ([]*bundleDataPart, error)
```

Comment:
Do you want to make read a private method?

Reply:
The initial thought was to only allow BundleDataSource implementations inside the charms package (this is also why bundleDataPart is un-exported). Given that we probably need to add a BundleDataSource implementation for dealing with remote bundles into another bundle I will make these public.